### PR TITLE
Added option hostname to docker service create. This option will allo…

### PR DESCRIPTION
**Pull request by vasil-yordanov**
This pull request is related to issue docker/docker#24877

**\- What I did** 
Added option hostname in docker service create command.

**\- How I did it**
I changed the docker client, updated the ContainerSpec structures in docker/swarm repo, so that this option is serialized to the swarm agents from the clients.

**\- How to verify it**
docker service create --name test --hostname vasko <image_id>
docker exec test.1.<task_id> hostname

**\- Description for the changelog**
This pull request will allow to specify hostname option to the docker service create command.
The same hostname will be set on all containers.

Signed-off-by: Vasil Yordanov vasil.yordanov@gmail.com

### CODE REVIEW

**Comment by vasil-yordanov** [comment]
This pull request is failing on build because it depends on the following pull requests: https://github.com/docker/engine-api/pull/352 
https://github.com/docker/swarmkit/pull/1318

**Comment by vasil-yordanov** [comment]
I could fix the failing-ci if I change hack/vendor.sh to download the swarmkik and engine-api dependencies from my forked versions of these repos. How to proceed? Should I wait for the for the engine-api and swarmkit pull requests to be accepted, or to change the vendor.sh?

**Comment by vdemeester** [comment]
@vasil-yordanov only the `vendor` *check* should fail because of vendoring (which is ok). The other ones fails because of `gofmt` :

```
16:24:08 ---> Making bundle: validate-gofmt (in bundles/1.13.0-dev/validate-gofmt)
16:24:08 These files are not properly gofmt'd:
16:24:08  - api/client/service/opts.go
16:24:08  - daemon/cluster/convert/container.go
16:24:08 
16:24:08 Please reformat the above files using "gofmt -s -w" and commit the result.
```

**Comment by vasil-yordanov** [comment]
@vdemeester I have fixed the formatting, but some check fails. I saw in the logs "Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#25437', new value: '#25437'". Is this some issue from build system side or there is something from my side that I have to do?

**Comment by thaJeztah** [comment]
@vasil-yordanov looks like the error is;

```
11:45:53 # github.com/docker/docker/api/client/service
11:45:53 api/client/service/opts.go:445: unknown swarm.ContainerSpec field 'Hostname' in struct literal
```

**Comment by vasil-yordanov** [comment]
@thaJeztah this error is because the code depends on github.com/docker/engine-api/types/swarm, which is in another github repo, where I have waiting pull request.
https://github.com/docker/engine-api/pull/352
Actually this pull request depend on pull requests on two other repos: docker/engine-api and docker/swamkit, I'm not sure if this is the same team that is responsible for accepting the pull request on these three repos or it is separated teams with good synchronization.
@vdemeester told me earlier that "only the vendor check should fail because of vendoring (which is ok).", so I believe that now this error is due to vendoring only.

**Comment by thaJeztah** [comment]
@vasil-yordanov what we usually do, is to temporarily include the changes in the vendored code in this PR (you can do separate commits for that); that way the code works, and CI can run. The vendor-CI will complain that you made changes to vendored code that don't match the commit/tag specified in `hack/vendor.sh`, but that's ok; once this PR is in code review, and the related PRs are merged, you can update the `hack/vendor.sh` and vendored code, so that the vendor CI goes green.

**Comment by vasil-yordanov** [comment]
Thank you @theJeztah for explaining me the process. I will commit vendor. Actually now I noticed that there is such folder in this repo. I was expecting that it is created by hach/vendor.sh during build time, and the dependencies are not kept here, but fetched when needed from the corresponding repo.

**Comment by vasil-yordanov** [comment]
@GordonTheTurtle I have checked that I have signed all my commits with git -s except the merges to the master of docker/docker repo which generates automatic commit message.
I also checked that "git rebase -i HEAD~5" does not show commits that I haven't signed, It do not show the merge commits.

These are my commits from "git reabase -i" command
pick 2a659dd Added option hostname to docker service create. This option will allow setting the same hostname on all containers spawned for given service
pick 3c4358d Reformated files with gofmt
pick 0009a68 Add vendor changes, to build the code and make CI to pass
In the git-rebase-todo I can not see merge commits.

All my commits have been signed when I did the commits, so it seems I do not need to amend the commits.

**Comment by justincormack** [comment]
@vasil-yordanov you should not have merge commits of master in your PR. You should remove the merge commit and rebase, and that will fix the signing issue.

**Comment by vasil-yordanov** [comment]
Does anyone has idea why this CI fails?
docker_cli_build_test.go is not related to my change for docker service create --hostname option.

**Comment by thaJeztah** [comment]
win2lin is currently having issues (there's a PR to fix that), not sure about windowsRS1, but we can restart. During design review, it's not a problem anyway

**Comment by vasil-yordanov** [comment]
@thaJeztah I have implemented the --hostname option to be a template, like --hostname=srv-*, in this case if we create a service with replicas>1 then the hostname will be srv-1, srv-2, ...
Only single replacement of * will happen. If there is not * at all then no replacement will be done and the hostname will be set the same on all containers for the given service.
Do you think this design is acceptable for you?
In my case I will not use * at all, I have implemented this because some people wanted it in docker/docker#24877

**Comment by vasil-yordanov** [comment]
I have removed my last commit after some discussion with stevvooe. "We need to be careful about creating a "template" system without thinking about consistent user experience."

**Comment by cpuguy83** [comment]
Not really sure why hostname would really matter in services (or even normal containers)... but I guess people find reasons.
In the linked issue it looks like there was still some discussion going on.
@mavenugo Are you ok with this?

**Comment by vasil-yordanov** [comment]
@stevvooe I start working on the adding hostname option in the docker repo, but do not know well the vendoring process. 
I changed the hack/vendor.sh to use my repo: `clone git github.com/vasil-yordanov/swarmkit.git`
Then I tried to build with make, but it fails due to missing field in swarm.ContainerSpec structure, because the vendoring didn't happen.
```
Building: bundles/1.13.0-dev/binary-client/docker-1.13.0-dev
# github.com/docker/docker/api/client/service
api/client/service/opts.go:447: unknown swarm.ContainerSpec field 'Hostname' in struct literal
Makefile:80: recipe for target 'binary' failed
make: *** [binary] Error 1
```
I tried to execute hack/vendor.sh manually before to build, but then error is:
```
api/client/service/update.go:19:2: cannot find package "github.com/flynn-archive/go-shlex" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/flynn-archive/go-shlex (vendor tree)
	/usr/local/go/src/github.com/flynn-archive/go-shlex (from $GOROOT)
	/go/src/github.com/flynn-archive/go-shlex (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/flynn-archive/go-shlex
cli/cobra.go:6:2: cannot find package "github.com/spf13/cobra" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra (vendor tree)
	/usr/local/go/src/github.com/spf13/cobra (from $GOROOT)
	/go/src/github.com/spf13/cobra (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/cobra
cli/flags/common.go:12:2: cannot find package "github.com/spf13/pflag" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/spf13/pflag (vendor tree)
	/usr/local/go/src/github.com/spf13/pflag (from $GOROOT)
	/go/src/github.com/spf13/pflag (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/pflag
Makefile:80: recipe for target 'binary' failed
make: *** [binary] Error 1
```
I could make the same changes I did before in docker/swarmkit repo by hand, based on the `clone git github.com/docker/swarmkit 8a761950fb4d9251c335dc6149a8a02756cb3b10`, but this seems a little bit strange to me, because it is error prone and double work.

Could you help me to understand the vendoring process?
How should I incorporate my changes in docker/swarm repo into the docker/docker repo?
Why you keep the vendor directory under VCS, after you have defined in vendor.sh the exact hashes and vendor repos from which the docker/docker code depends on?

**Comment by stevvooe** [comment]
> How should I incorporate my changes in docker/swarm repo into the docker/docker repo?

Just include the revision of swarmkit that has the `Hostname` changes. When you run the hack/vendor.sh script, it will update the dependencies locally. Confirm these are correct then add these changes to your PR.

> Why you keep the vendor directory under VCS, after you have defined in vendor.sh the exact hashes and vendor repos from which the docker/docker code depends on?

Downloading dependencies is error prone and time consuming. By including them directly in the project, you ensure that the project can be built, even if the upstream dependencies are unavailable or the version you were using is broken. In Go, this is a defacto standard approach. We retain the hashes to ensure we have provenance of which versions led to which updates.

**Comment by vasil-yordanov** [comment]
@stevvooe Thank you for clarifying me the docker vendoring process. 
I already tried to execute the ./hack/vendor.sh, but then the 'make' fails with the error I wrote you before:

```
---> Making bundle: binary (in bundles/1.13.0-dev/binary)
Building: bundles/1.13.0-dev/binary-client/docker-1.13.0-dev
api/client/service/update.go:19:2: cannot find package "github.com/flynn-archive/go-shlex" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/flynn-archive/go-shlex (vendor tree)
	/usr/local/go/src/github.com/flynn-archive/go-shlex (from $GOROOT)
	/go/src/github.com/flynn-archive/go-shlex (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/flynn-archive/go-shlex
cli/cobra.go:6:2: cannot find package "github.com/spf13/cobra" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra (vendor tree)
	/usr/local/go/src/github.com/spf13/cobra (from $GOROOT)
	/go/src/github.com/spf13/cobra (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/cobra
cli/flags/common.go:12:2: cannot find package "github.com/spf13/pflag" in any of:
	/go/src/github.com/docker/docker/vendor/github.com/spf13/pflag (vendor tree)
	/usr/local/go/src/github.com/spf13/pflag (from $GOROOT)
	/go/src/github.com/spf13/pflag (from $GOPATH)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/pflag
Makefile:80: recipe for target 'binary' failed
make: *** [binary] Error 1
```

**The same error occurs when I do:**
```
git clone https://github.com/docker/docker.git
./hack/vendor.sh
make
```

In the same time when I do git diff after execution of ./hack/vendor.sh, it shows me a lot of untracked files and files not staged for commit in 'vendor'  folder.

**Comment by stevvooe** [comment]
What is the value of `GOPATH` when you run the command?

> In the same time when I do git diff after execution of ./hack/vendor.sh, it shows me a lot of untracked files and files not staged for commit in 'vendor' folder.

This is normal. You'll have to check these. I'd recommend doing this in a separate commit.

**Comment by vasil-yordanov** [comment]
@stevoo Sorry for the inconvenience. I just noticed that ./hack/vendor.sh was not executed successful, due to missing 'hg'. I've missed this message.

`bitbucket.org/ww/goautoneg @ 75cd24fc2f2c2a2088577d12123ddee5f54e0675: clone, hack/.vendor-helpers.sh: line 44: hg: command not found`

After installing the mercurial now everything is working.
I pushed my re-vendored fork, but will test if this hostname option is still working and will let you know tomorrow.

**Comment by mostolog** [comment]
@cpuguy83 
> Not really sure why hostname would really matter in services (or even normal containers)... but I guess people find reasons.

On the one hand, it's easier to know where you are running commands when the prompt shows the task.slot as hostname. Definitely _root@2d425a02daf1_ doesn't help.
On the other hand, if _--hostname_ is somehow linked to DNS name of the running container, containers could find each other by name, instead of IP, being easier to handle IP changes.

I'll love each instance to be named as <task>.<slot> like @thaJeztah said at https://github.com/docker/docker/pull/24973#issuecomment-234904529

Regards

**Comment by vasil-yordanov** [comment]
@stevvooe I have re-vendored this PR and also tested manually the new feature with hostanme, and it is working. Could you make me code review and let me know how to proceed?

**Comment by stevvooe** [comment]
@vasil-yordanov Thanks for getting this into a workable state!

@mrjana @aluzzardi @crosbymichael @mavenugo Could you comment on the problems with allowing setting the hostname of all instances of a service and possible future implications? Are there areas that need to be limited to allow for future hostname oriented features?

**Comment by crosbymichael** [comment]
@stevvooe i cannot think of any on the system side of things.  Its fine to reuse the hostname and does not cause any conflicts.

**Comment by mostolog** [comment]
@crosbymichael @stevvooe IMHO If all containers share the same hostname, there could be some scenarios where problems could arise. Let me give you a (_pretended_) example:

Redis clusters use a 1-16383 slot array, in which keys are stored based on hash(key), hence **each node is responsible for a few slots**. Now consider node-1 handles slot 1 and node-2 handles slot 2.

Redis client can request node-1 to put data on slot 1 and that's ok, but if also request to put data on slot 2, node-1 will response "MOVED to node-2". (To be honest: MOVED to IP:PORT, but **using hostnames ilustrates the problem**)

It's also recommended to have _smart_ Redis client, which remember the node-slot topology, avoiding MOVED responses and improving performance. In the case of sharing hostnames, that could also be a problem:
```
 slot 1: mynode # instead of node-1
 slot 2: mynode # instead of node-2
```
[More information about Redis clusters](http://redis.io/topics/cluster-spec)

To sum up: IMHO each hostname should be the same as the service short name(without hash).
ie: node-1, node-2, even NOT USING DOTS "." for _task.slot_ naming, but _task-slot_.
[I suggest reading](http://stackoverflow.com/a/2183140)

Hope it helps.

**Comment by stevvooe** [comment]
@mostolog I think you're confusing the hostname seen within the container process with the hostname exported to the network. As it is planned, this hostname parameter would affect the hostname as seen by container process. Much of what you're discussing is proposed in https://github.com/docker/swarmkit/issues/1242.

The main issue is whether or not bind these two concepts. While what you are demonstrating is important, it is largely orthogonal to this discussion. In fact, showing that these are easily confused is an argument on the side of considering this parameter in the context of a DNS mapping scheme.

We still need to hear from @mavenugo @mrjana on this topic!

**Comment by mostolog** [comment]
> @stevvooe 
> I think you're confusing the hostname seen within the container process with the hostname exported to the network. 

Indeed, I though they _should be_ the same!

> The main issue is whether or not bind these two concepts. While what you are demonstrating is important, it is largely orthogonal to this discussion. In fact, showing that these are easily confused is an argument on the side of considering this parameter in the context of a DNS mapping scheme.

Totally agree.
Going docker/swarmkit#1242

**Comment by stevvooe** [comment]
> Indeed, I though they should be the same!

@mostolog If only life were so simple!!!

Thanks for the feedback! Let's keep bouncing this around so we can get this right.

**Comment by stevvooe** [comment]
@thaJeztah We merged the functionality in swarmkit. This PR will likely need to be carried for integration into docker.

**Comment by niau** [comment]
@stevvooe so since https://github.com/docker/swarmkit/pull/1688 was merges I assume we have to also do our job and merge that one in vendor. Or since actually @vasil-yordanov has referenced his repo in vendor.sh we don't need to do stuff here and that one can also be closed?

**Comment by stevvooe** [comment]
@niau That is the process. Let us know if you need help!

**Comment by niau** [comment]
uha @stevvooe  I assume that I had really lagged  the dogs here. 
This is basically the result of updating vendor.sh, executing it, and committing the various changes to vendor folders. It was a strange that the new run of vendor.sh caused also a lot of changes not only to swarmkit but also to other vendor folders. Is this how it is and should I commit also those? If not the changes  are in separate commits to facilitate easy reverting.

**Comment by stevvooe** [comment]
@niau You can do something like `./hack/vendor.sh github.com/docker/swarmkit` to only update that dependency. Let's keep the updates scoped to swarmkit.

**Comment by thaJeztah** [comment]
Ok, this PR wasn't actually merged, but possibly @vasil-yordanov rebased/updated his master, and GitHub shows the exact same tree as https://github.com/docker/docker/commit/80d6d2e129805e8ee9257641b922e47a01f16736, so automatically marks it as "merged", even though nothing was actually merged;

screen shot 2016-10-26 at 5 22 56 pm

**Comment by niau** [comment]
@thaJeztah  very true I simply rebased it.

**Comment by thaJeztah** [comment]
@niau looks like something went wrong during the rebase, because the change set in this PR is not empty https://github.com/docker/docker/pull/25437/files

**Comment by niau** [comment]
Thanks @theJeztah I assume so. We are missing one commit. I have recommited it in vasil's repo, but it does not appear here. It looks like the only option is a new PR.

**Comment by vdemeester** [comment]
@niau yep, and it would be best to open the PR using a branch (that way, if you change/rebase master, it shouldn't happen again 👼)

**Comment by niau** [comment]
@vdemeester @thaJeztah sorry for all the noise I hope that the latest PL is exactly as it should be. It is from a branch, it is rebased and hopefully all tests pass :)
